fdbserver/RestoreMaster.actor.cpp

```diff
@@ -0,0 +1,1339 @@
+	state std::map<UID, RestoreLoaderInterface>::iterator loader;
+	for (loader = self->loadersInterf.begin(); loader != self->loadersInterf.end(); loader++) {
+		self->cmdID.nextCmd();
+		loop {
+			try {
+				cmdReplies.clear();
+				printf("[CMD] Node:%s Notify node:%s about appliers key range\n", self->describeNode().c_str(), loader->first.toString().c_str());
+				cmdReplies.push_back( loader->second.setApplierKeyRangeVectorRequest.getReply(RestoreSetApplierKeyRangeVectorRequest(self->cmdID, appliers, ranges)) );
+				printf("[INFO] Wait for node:%s to accept the cmd Notify_Loader_ApplierKeyRange\n", loader->first.toString().c_str());
+				std::vector<RestoreCommonReply> reps = wait( timeoutError( getAll(cmdReplies), FastRestore_Failure_Timeout ) );
+				printf("Finished Notify_Loader_ApplierKeyRange: number of appliers:%d\n", appliers.size());
+				cmdReplies.clear();
+				break;
+			} catch (Error &e) {
+				fprintf(stdout, "[ERROR] Node:%s, Commands before cmdID:%s timeout\n", self->describeNode().c_str(), self->cmdID.toString().c_str());
+			}
+		}
+	}
+
+	return Void();
+}
+
+
+ACTOR static Future<Void> finishRestore(Reference<RestoreMasterData> self, Database cx, Standalone<VectorRef<RestoreRequest>> restoreRequests) {
+	// Make restore workers quit
+	state std::vector<Future<RestoreCommonReply>> cmdReplies;
+	state std::map<UID, RestoreLoaderInterface>::iterator loader;
+	state std::map<UID, RestoreApplierInterface>::iterator applier;
+	loop {
+		try {
+			cmdReplies.clear();
+			self->cmdID.initPhase(RestoreCommandEnum::Finish_Restore);
+
+			for ( loader = self->loadersInterf.begin(); loader != self->loadersInterf.end(); loader++ ) {
+				self->cmdID.nextCmd();
+				cmdReplies.push_back(loader->second.finishRestore.getReply(RestoreSimpleRequest(self->cmdID)));
+			}
+			for ( applier = self->appliersInterf.begin(); applier != self->appliersInterf.end(); applier++ ) {
+				self->cmdID.nextCmd();
+				cmdReplies.push_back(applier->second.finishRestore.getReply(RestoreSimpleRequest(self->cmdID)));
+			}
+
+			if (!cmdReplies.empty()) {
+				std::vector<RestoreCommonReply> reps =  wait( timeoutError( getAll(cmdReplies), FastRestore_Failure_Timeout / 100 ) );
+				//std::vector<RestoreCommonReply> reps =  wait( getAll(cmdReplies) );
+				cmdReplies.clear();
+			}
+			dbprintf("All restore workers have quited\n");
+
+			break;
+		} catch(Error &e) {
+			printf("[ERROR] At sending finishRestore request. error code:%d message:%s. Retry...\n", e.code(), e.what());
+			self->loadersInterf.clear();
+			self->appliersInterf.clear();
+			cmdReplies.clear();
+			wait( _collectRestoreRoleInterfaces(self, cx) );
+		}
+	}
+
+	// Notify tester that the restore has finished
+	state ReadYourWritesTransaction tr3(cx);
+	loop {
+		try {
+			tr3.reset();
+			tr3.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr3.setOption(FDBTransactionOptions::LOCK_AWARE);
+			tr3.clear(restoreRequestTriggerKey);
+			tr3.clear(restoreRequestKeys);
```

Comment:
If any restore request came in after the `getRange(restoreRequestKeys)` to collect the restore requests, then this would remove them also.